include/revng/Model/FunctionEdge.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "revng/Model/FunctionEdgeType.h"
+#include "revng/Model/GeneratedFunctionEdge.h"
+
+/// An edge on the CFG
+class model::FunctionEdge : public model::generated::FunctionEdge {
+public:
+  using generated::FunctionEdge::FunctionEdge;
+
+public:
+  bool operator<(const FunctionEdge &Other) const {
+    auto ThisTie = std::tie(Destination, Type);
+    auto OtherTie = std::tie(Other.Destination, Other.Type);
+    return ThisTie < OtherTie;
+  }
+
+public:
+  static bool classof(const FunctionEdge *A) {
+    return not FunctionEdgeType::isCall(A->Type);
+  }
+
+  bool verify() const debug_function;
+  bool verify(bool Assert) const debug_function;
+  bool verify(VerifyHelper &VH) const;
+  void dump() const debug_function;
+
+public:
+  // TODO: ideally we should not need to define this here
+  static constexpr const char *Tag = "!FunctionEdge";
+};
+
+// This must be included first
+#include "revng/Model/GeneratedFunctionEdge_late.h"
+// And this second. The comments don't let clang-format reorder them.
+#include "revng/Model/CallEdge.h"
+
+template<>
+struct KeyedObjectTraits<UpcastablePointer<model::FunctionEdge>> {
+  using Key = model::FunctionEdge::Key;
+  static Key key(const UpcastablePointer<model::FunctionEdge> &Obj) {
+    return { Obj->Destination, Obj->Type };
+  }
+
+  static UpcastablePointer<model::FunctionEdge> fromKey(const Key &Obj) {
+    using namespace model;
+    using ResultType = UpcastablePointer<FunctionEdge>;
+    if (FunctionEdgeType::isCall(std::get<1>(Obj))) {
+      return ResultType(new CallEdge(std::get<0>(Obj), std::get<1>(Obj)));
+    } else {
+      return ResultType(new FunctionEdge(std::get<0>(Obj), std::get<1>(Obj)));
+    }
+  }
+};
+
+template<>
+class llvm::yaml::MappingTraits<UpcastablePointer<model::FunctionEdge>>
+  : public PolymorphicMappingTraits<UpcastablePointer<model::FunctionEdge>> {};
```

Comment:
I'd like this to be generated too.

Reply:
Is it possible to generate `fromKey`? I guess I could instanciate a `FunctionEdge` from the key and then see if it is a `CallEdge` or a `FunctionEdge` using `classof`

Reply:
Yeah, `classof` enables you to decouple from `isCall`.

Reply:
Done